### PR TITLE
net/mqtt: Enable TLS support

### PR DESCRIPTION
**Pull request by askawu**
1. Convert MQTT lib to use net app API and modify MQTT related tests
2. Enable TLS support of MQTT lib

TODO:
1. Handle PSK and CA for mqtt_publisher
2. Add test cases for MQTT with TLS

### CODE REVIEW

File: samples/net/mqtt_publisher/prj_frdm_k64f.conf [hunk 1]
```diff
@@ -25,6 +25,7 @@ CONFIG_NET_IPV6=y
 CONFIG_MQTT_LIB=y
 
 CONFIG_NET_APP_SETTINGS=y
+CONFIG_NET_APP_CLIENT=y
```
Comment (jukkar) [inline samples/net/mqtt_publisher/prj_frdm_k64f.conf]: This is not needed as you have that automatically selected when CONFIG_MQTT_LIB is enabled.

**Comment by askawu** [comment]
@pfalcon @jukkar please help to review my patch

File: samples/net/mqtt_publisher/src/main.c [hunk 2]
```diff
@@ -420,57 +393,32 @@ static int network_setup(struct net_context **net_ctx, const char *local_addr,
...
-		printk("set_addr (server) error\n");
-		goto lb_exit;
-	}
-
-	rc = net_context_connect(*net_ctx, &server_sock, addr_len, NULL,
-				 APP_SLEEP_MSECS, NULL);
-	if (rc) {
-		printk("net_context_connect error\n");
-		goto lb_exit;
+	p = net_if_ipv4_addr_add(net_if_get_default(),
```
Comment (jukkar) [inline samples/net/mqtt_publisher/src/main.c]: If you call net_app_init(), then that function will make sure that the IP addresses are set to the network interface and you can remove this and the corresponding ipv6 call from the app.

File: tests/net/lib/mqtt_publisher/src/test_mqtt_publish.c [hunk 3]
```diff
@@ -327,96 +326,42 @@ static int test_disconnect(void)
...
-	if (rc) {
-		TC_PRINT("net_context_bind error\n");
 		goto lb_exit;
 	}
-
-	rc = set_addr(&server_sock, server_addr, server_port);
-	if (rc) {
-		TC_PRINT("set_addr (server) error\n");
+#else
+	rc = net_addr_pton(AF_INET6, CONFIG_NET_APP_MY_IPV4_ADDR, &addr);
```
Comment (jukkar) [inline tests/net/lib/mqtt_publisher/src/test_mqtt_publish.c]: There should be no need to set the IP addresses manually, just call net_app_init() and it will handle all of this setup.

File: tests/net/lib/mqtt_subscriber/src/test_mqtt_subscribe.c [hunk 4]
```diff
@@ -364,97 +369,43 @@ static int test_disconnect(void)
...
+	if (rc < 0) {
 		goto lb_exit;
 	}
-
-	rc = net_context_connect(*net_ctx, &server_sock, addr_len, NULL,
-			APP_SLEEP_MSECS, NULL);
-	if (rc) {
-		printk("net_context_connect error\n"
-				"Is the server (broker) up and running?\n");
+	p = net_if_ipv4_addr_add(net_if_get_default(),
```
Comment (jukkar) [inline tests/net/lib/mqtt_subscriber/src/test_mqtt_subscribe.c]: same comment to this too, use net_app_init() here

File: ext/lib/crypto/mbedtls/configs/config-mini-tls1_2.h [hunk 5]
```diff
@@ -74,7 +74,7 @@
 #define MBEDTLS_SSL_ALL_ALERT_MESSAGES
 #endif
 
-#define MBEDTLS_SSL_MAX_CONTENT_LEN  1500
+#define MBEDTLS_SSL_MAX_CONTENT_LEN  2500
```
Comment (jukkar) [inline ext/lib/crypto/mbedtls/configs/config-mini-tls1_2.h]: Please put this into separate patch. Even better would be to add if defined(MQTT...) here if this increase is required by MQTT only.

File: samples/net/mqtt_publisher/src/main.c [hunk 6]
```diff
@@ -61,15 +61,41 @@ struct mqtt_client_ctx {
...
-			 const char *server_addr, u16_t server_port);
+static int network_setup(void);
+
+#if defined(CONFIG_MQTT_LIB_TLS)
+/* TLS */
+#define TLS_SNI_HOSTNAME "localhost"
+#define TLS_REQUEST_BUF_SIZE 1500
+#define TLS_PRIVATE_DATA "Zephyr TLS mqtt publisher"
+
+#if !defined(CONFIG_NET_APP_TLS_STACK_SIZE)
```
Comment (jukkar) [inline samples/net/mqtt_publisher/src/main.c]: The TLS stack size value is set by default so no need to set it here too. If the default stack size is too long, then its value can be tweaked in prj conf file. I noticed that I had that setup like this in echo-client too but I will need to change that.

File: samples/net/mqtt_publisher/src/main.c [hunk 7]
```diff
@@ -61,15 +61,41 @@ struct mqtt_client_ctx {
...
+static u8_t tls_request_buf[TLS_REQUEST_BUF_SIZE];
+
+NET_STACK_DEFINE("mqtt_tls_stack", tls_stack,
+		CONFIG_NET_APP_TLS_STACK_SIZE, CONFIG_NET_APP_TLS_STACK_SIZE);
+
+NET_APP_TLS_POOL_DEFINE(tls_mem_pool, 30);
+
+int setup_cert(struct net_app_ctx *ctx, void *cert)
+{
+	/* TODO: PSK and CA verification */
```
Comment (jukkar) [inline samples/net/mqtt_publisher/src/main.c]: Before TLS support patch can be accepted, this part of the code needs to be implemented too.

File: samples/net/mqtt_publisher/src/main.c [hunk 8]
```diff
@@ -243,45 +269,38 @@ static void publisher(void)
...
+#if defined(CONFIG_MQTT_LIB_TLS)
+	/** TLS setup */
+	client_ctx.mqtt_ctx.request_buf = tls_request_buf;
+	client_ctx.mqtt_ctx.request_buf_len = TLS_REQUEST_BUF_SIZE;
+	client_ctx.mqtt_ctx.personalization_data = TLS_PRIVATE_DATA;
+	client_ctx.mqtt_ctx.personalization_data_len = strlen(TLS_PRIVATE_DATA);
+	client_ctx.mqtt_ctx.cert_host = TLS_SNI_HOSTNAME;
+	client_ctx.mqtt_ctx.tls_mem_pool = &tls_mem_pool;
+	client_ctx.mqtt_ctx.tls_stack = tls_stack;
+	client_ctx.mqtt_ctx.tls_stack_size = CONFIG_NET_APP_TLS_STACK_SIZE;
```
Comment (jukkar) [inline samples/net/mqtt_publisher/src/main.c]: You need to use K_THREAD_STACK_SIZEOF() macro here like this

...tls_stack_size = K_THREAD_STACK_SIZEOF(tls_stack);

**Review by jukkar — CHANGES_REQUESTED** [review]
Thanks for submitting this. I had this in my todo list but fortunately you got there first.

**Comment by askawu** [comment]
New patches are submitted
1. Remove network setup code since it will be done by net_app_init()
2. Use K_THREAD_STACK_SIZEOF() for the stack size
3. Handle PSK and CA verification. Also, I found a bug of mosquitto that PSK with leading zero will result TLS handshake failure.  See https://github.com/eclipse/mosquitto/pull/498 for more info.
4. Add a new changes for mbedtls config

**Comment by askawu** [comment]
Update:
Add a "build-only" test case to ensure there's no build error if CONFIG_MQTT_LIB_TLS is enabled.

**Review by jukkar — COMMENTED** [review]
Looks quite good, minor comment thou.

File: samples/net/mqtt_publisher/src/main.c [hunk 9]
```diff
@@ -61,15 +61,69 @@ struct mqtt_client_ctx {
...
-			 const char *server_addr, u16_t server_port);
+static int network_setup(void);
+
+#if defined(CONFIG_MQTT_LIB_TLS)
+
+#include "test_certs.h"
+
+/* TLS */
+#define TLS_SNI_HOSTNAME "localhost"
+#define TLS_REQUEST_BUF_SIZE 1500
```
Comment (jukkar) [inline samples/net/mqtt_publisher/src/main.c]: So I was just wondering here why would we need to have 2500 byte long mbedtls setting if we always read max 1500 bytes long data (size of the request buf).

Comment (askawu) [inline samples/net/mqtt_publisher/src/main.c]: Good question, I will figure it out. Thanks

Comment (askawu) [inline samples/net/mqtt_publisher/src/main.c]: I think MBEDTLS_SSL_MAX_CONTENT_LEN is related to the maximum size of SSL record, According to the SSL protocol, it can be 16384 at most. During the handshake with mosquitto, the SSL record is larger than 1500, this is why handshake failed.

TLS_REQUEST_BUF_SIZE is the buffer size we used to hold the data for mbedtls_ssl_read() and mbedtls_ssl_write(). But it should be less than MBEDTLS_SSL_MAX_CONTENT_LEN due to the protocol limitation.

**Comment by askawu** [comment]
Rebased on latest upstream/master

**Review by jukkar — APPROVED** [review]
LGTM
